com.unity.netcode.gameobjects/Runtime/Core/NetworkBehaviour.cs

```diff
@@ -235,14 +235,42 @@ internal void __endSendClientRpc(ref FastBufferWriter bufferWriter, uint rpcMeth
 #if DEVELOPMENT_BUILD || UNITY_EDITOR
             if (NetworkManager.__rpc_name_table.TryGetValue(rpcMethodId, out var rpcMethodName))
             {
-                foreach (var client in NetworkManager.ConnectedClients)
+                if (clientRpcParams.Send.TargetClientIds != null)
                 {
-                    NetworkManager.NetworkMetrics.TrackRpcSent(
-                        client.Key,
-                        NetworkObject,
-                        rpcMethodName,
-                        __getTypeName(),
-                        rpcWriteSize);
+                    foreach (var targetClientId in clientRpcParams.Send.TargetClientIds)
+                    {
+                        NetworkManager.NetworkMetrics.TrackRpcSent(
+                            targetClientId,
+                            NetworkObject,
+                            rpcMethodName,
+                            __getTypeName(),
+                            rpcWriteSize);
+                    }
+                }
+                else if (clientRpcParams.Send.TargetClientIdsNativeArray != null)
+                {
+                    foreach (var targetClientId in clientRpcParams.Send.TargetClientIdsNativeArray)
+                    {
+                        NetworkManager.NetworkMetrics.TrackRpcSent(
+                            targetClientId,
+                            NetworkObject,
+                            rpcMethodName,
+                            __getTypeName(),
+                            rpcWriteSize);
+                    }
+                }
+                else
+                {
+                    var observerEnumerator = NetworkObject.Observers.GetEnumerator();
+                    while (observerEnumerator.MoveNext())
+                    {
+                        NetworkManager.NetworkMetrics.TrackRpcSent(
+                            observerEnumerator.Current,
+                            NetworkObject,
+                            rpcMethodName,
+                            __getTypeName(),
+                            rpcWriteSize);
+                    }
```

Comment:
If the `clientRpcParams.Send.TargetClientIdsNativeArray` didn't cause Boxing allocation I would suggest using something like:
```
IEnumerable<ulong> enumerable;

if (clientRpcParams.Send.TargetClientIds != null)
{
    enumerable = clientRpcParams.Send.TargetClientIds;
}
else if (clientRpcParams.Send.TargetClientIdsNativeArray != null)
{
    enumerable = clientRpcParams.Send.TargetClientIdsNativeArray;
}
else
{
    enumerable = NetworkObject.Observers;
}

foreach (var networkObject in enumerable)
{
    NetworkManager.NetworkMetrics.TrackRpcSent(
        networkObject,
        NetworkObject,
        methodName,
        __getTypeName(),
        rpcWriteSize);
}
```

But in this case, it seems there's no other way to achieve the most efficient code without sacrificing a bit of readability.

So I believe your code is the best choice here.

Reply:
Yeah, that was my thought, too.